types/queries_test.go

```diff
@@ -61,3 +61,51 @@ func TestValidatorWithData(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, reval, val)
 }
+
+func TestQueryResponseWithEmptyData(t *testing.T) {
+	cases := map[string]struct{
+		req QueryResponse
+		resp string
+		unmarshal bool
+	}{
+		"ok with data": {
+			req: QueryResponse{Ok:  []byte("foo")},
+			// base64-encoded "foo"
+			resp: `{"ok":"Zm9v"}`,
+			unmarshal: true,
+		},
+		"error": {
+			req: QueryResponse{Err: "try again later"},
+			resp: `{"error":"try again later"}`,
+			unmarshal: true,
+		},
+		"ok with empty slice": {
+			req: QueryResponse{Ok: []byte{}},
+			resp: `{"ok":""}`,
+			unmarshal: true,
+		},
+		"nil data": {
+			req: QueryResponse{},
+			resp: `{"ok":""}`,
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			data, err := json.Marshal(tc.req)
+			require.NoError(t, err)
+			require.Equal(t, tc.resp, string(data))
+
+			// if unmarshall, make sure this comes back to the proper state
+			if tc.unmarshal {
+				var parsed QueryResponse
+				err = json.Unmarshal(data, &parsed)
```

Comment:
Do we even need the ability to deserialize `QueryResponse`? If I understand correctly, those are always created in Go and then serialized for the VM.

Reply:
We do, in this test:

```
	// call query which will call canonicalize address
	badApi := NewMockFailureAPI()
	gasMeter3 := NewMockGasMeter(100000000)
	query := []byte(`{"verifier":{}}`)
	igasMeter3 := GasMeter(gasMeter3)
	res, _, err := Query(cache, id, env, query, &igasMeter3, store, badApi, &querier, 100000000)
	require.NoError(t, err)
	var resp types.QueryResponse
	err = json.Unmarshal(res, &resp)
	require.NoError(t, err)

	// ensure the error message is what we expect (system -ok, stderr -generic)
	require.Nil(t, resp.Ok)
	// with this error
	require.Equal(t, "Generic error: humanize_address errored: mock failure - human_address", resp.Err)
```

Reply:
I just wanted a test that I didn't break it.
Of course I cannot test that it unmarshals both as `nil` and `[]byte{}` just one, which is why i skip that test one place.